File: src/main/scala/Day2/Day2.scala
package Day2

import scala.io.Source

object Day2 {

    def main(args: Array[String]): Unit = {
        val source = Source.fromFile("src/main/scala/Day2/input")
        val input = source.getLines().toList
        source.close()

        println(part1(input) + " (correct: 2322630)")
        println(part2(input) + " (correct: 2105273490)")
    }

    def part1(xs: List[String]): Int = {
        var forward = 0
        var down = 0
        xs.foreach(f = e => e.split(" ").head match {
            case "forward" => forward += e.split(" ")(1).toInt
            case "up" => down -= e.split(" ")(1).toInt
            case "down" => down += e.split(" ")(1).toInt
        })
        forward * down
    }

    def part2(xs: List[String]): Int = {
        var forward = 0
        var down = 0
        var aim = 0
        xs.foreach(e => e.split(" ").head match {
            case "forward" =>
                forward += e.split(" ")(1).toInt
                down += aim * e.split(" ")(1).toInt
            case "up" => aim -= e.split(" ")(1).toInt
            case "down" => aim += e.split(" ")(1).toInt
        })
        forward * down
    }
}

File: src/main/scala/Day2/Day2.hs
data Instruction = Forward Int | Up Int | Down Int deriving Show
type Position = (Int, Int, Int) {- Forward, Depth, Aim (for part2) -}

main = do
  ls <- fmap lines (readFile "Day2/input")
  let instructionList = map (parseInstruction . makeTuple) ls
  print $ part1 instructionList
  print $ part2 instructionList


part1 :: [Instruction] -> Int
part1 instruction = forward * down where (forward, down, aim) = foldl updatePart1 (0,0,0) instruction

part2 :: [Instruction] -> Int
part2 instruction = forward * down where (forward, down, aim) = foldl updatePart2 (0,0,0) instruction

updatePart1 :: Position -> Instruction -> Position
updatePart1 (x, y, z) (Forward n) = (x + n, y, z)
updatePart1 (x, y, z) (Down    n) = (x, y + n, z)
updatePart1 (x, y, z) (Up      n) = (x, y - n, z)

updatePart2 :: Position -> Instruction -> Position
updatePart2 (x, y, z) (Forward n) = (x + n, y + z * n, z)
updatePart2 (x, y, z) (Down    n) = (x, y, z + n)
updatePart2 (x, y, z) (Up      n) = (x, y, z - n)

makeTuple :: String -> (String, Int)
makeTuple s = (head str, read $ last str :: Int) where str = words s

parseInstruction :: (String, Int) -> Instruction
parseInstruction (dir, val)
    | dir == "forward" = Forward(val)
    | dir == "up"      = Up(val)
    | dir == "down"    = Down(val)
    | otherwise        = error "WTF happened?"